src/install/dracut-install.c

```diff
@@ -1515,7 +1515,114 @@ static bool check_module_path(const char *path)
         return true;
 }
 
-static int install_dependent_module(struct kmod_module *mod, int *err)
+static int find_module_from_sysfs_node(struct kmod_ctx *ctx, const char *sysfs_node, int sysfs_node_len,
+                                       struct kmod_list **modules)
+{
+        char modalias_path[PATH_MAX] = {0};
+        snprintf(modalias_path, sizeof(modalias_path), "%.*s/modalias", sysfs_node_len, sysfs_node);
+
+        _cleanup_close_ int modalias_file = -1;
+        if ((modalias_file = open(modalias_path, O_RDONLY | O_CLOEXEC)) == -1)
+                return 0;
+
+        char alias[getpagesize()];
+        ssize_t len = read(modalias_file, alias, sizeof(alias));
+        alias[len - 1] = '\0';
+
+        return kmod_module_new_from_lookup(ctx, alias, modules);
+}
+
+static int find_suppliers_for_sys_node(struct kmod_ctx *ctx, Hashmap *suppliers, const char *node_path_raw,
+                                       size_t node_path_len)
+{
+        char node_path[PATH_MAX] = {};
+        memcpy(node_path, node_path_raw, node_path_len);
+        DIR *d;
+        struct dirent *dir;
+        d = opendir(node_path);
+        if (d) {
+                while ((dir = readdir(d)) != NULL) {
+                        if (strstr(dir->d_name, "supplier") != NULL) {
+                                char supplier_path[PATH_MAX] = {0};
+                                if (realpath(dir->d_name, supplier_path)) {
+                                        char supplier_slice[] = "/supplier";
+
+                                        if (strlen(supplier_path) > sizeof(supplier_path) - strlen(supplier_slice) - 1)
+                                                return -1;
+
+                                        strncat(supplier_path, supplier_slice, sizeof(supplier_path) - strlen(supplier_path) - 1);
+
+                                        _cleanup_kmod_module_unref_list_ struct kmod_list *list = NULL;
+                                        struct kmod_list *l = NULL;
+
+                                        if (find_module_from_sysfs_node(ctx, supplier_path, strlen(supplier_path), &list) >= 0) {
+                                                kmod_list_foreach(l, list) {
+                                                        struct kmod_module *mod = kmod_module_get_module(l);
+                                                        char *supplier_module = strdup(kmod_module_get_name(mod));
+                                                        kmod_module_unref(mod);
+
+                                                        int err = hashmap_put(suppliers, supplier_module, supplier_module);
+                                                        if (err < 0) {
+                                                                free(supplier_module);
+                                                                return err;
+                                                        }
```

Comment:
I don't think you need to `return err` here – hashmap_put() returns -EEXIST if `supplier_module` exists in the map already (the other error is ENOMEM but dracut-install doesn't care about allocations failing anyway) so just reduce to `if (hashmap_put()) free(supplier_module);` or equivalent and it's fine.